SimG4CMS/Calo/src/CaloSD.cc

```diff
@@ -458,111 +468,91 @@ std::string CaloSD::printableDecayChain(const std::vector<unsigned int>& decayCh
+    edm::LogVerbatim("DoFineCalo") << "Track " << id << " parent already cached: " << it->second;
 #endif
-    hitID.setFineTrackID(currentTrack->GetTrackID());
-    hit->setID(hitID);  // Actually overwrite the ID for the hit
-    trkInfo->storeTrack(true);
-    return;
+    return it->second;
   }
-  // currentTrack itself does not pass thresholds / does not cross boundary; go through its history to find a track that does
-  TrackWithHistory* recordTrackWithHistory;
-  // Keep track of decay chain of this track for debugging purposes
-  std::vector<unsigned int> decayChain;
-  decayChain.push_back(currentTrack->GetTrackID());
-  // Find the first parent of this track that passes the required criteria
-  // Start from first parent
-  unsigned int recordTrackID = currentTrack->GetParentID();
+  // Then see if the track itself crosses the boundary
+  else if (trkInfo->crossedBoundary()) {
 #ifdef EDM_ML_DEBUG
-  edm::LogVerbatim("DoFineCalo") << "Trying to find the first parent of hit " << hit->getUnitID()
-                                 << " that passes saving criterion (crosses boundary or specific criterion)"
-                                 << "; starting with first parent track " << recordTrackID;
+    edm::LogVerbatim("DoFineCalo") << "Track " << id << " crosses boundary itself";
 #endif
-  // Check whether this first parent actually exists
-  if (recordTrackID <= 0) {
-    // Track ID 0 is not a track;
-    // This means the current currentTrack has no parent, but apparently it also didn't fit saving criteria
-    throw cms::Exception("Unknown", "CaloSD") << "ERROR: Track " << currentTrack->GetTrackID()
-                                              << " has no parent, does not fit saving criteria, but left hit "
-                                              << hit->getUnitID() << "; recording it but it's weird!";
+    boundaryCrossingParentMap_[id] = id;
+    trkInfo->storeTrack(true);
+    return id;
   }
-  // Start progressing through the track's history
+  // Else, traverse the history of the track
+  std::vector<unsigned int> decayChain{id};
+#ifdef EDM_ML_DEBUG
+  edm::LogVerbatim("DoFineCalo") << "Track " << id << ": Traversing history to find boundary-crossing parent";
+#endif
+  unsigned int parentID = track->GetParentID();
   while (true) {
-    // Record the decay chain for debugging purposes
-    decayChain.push_back(recordTrackID);
-    recordTrackWithHistory = m_trackManager->getTrackByID(recordTrackID);
-    if (recordTrackID < (unsigned int)hitID.trackID()) {
-      // A parent of the currentTrack has a lower trackID than the current
-      // hitID.trackID(). This means the current hitID.trackID() does not point
-      // to the earliest ancestor of the currentTrack.
-      // The current hitID.trackID() might not be a saved track yet, but the
-      // ancestor is *always* a saved track.
-      // Fix this by overwriting the hitID's track ID.
+    if (parentID == 0)
```

Comment:
@tklijnsma , parentID may be 0 for primary tracks.

Reply:
@civanch This loop searches for the ancestor of a track that crosses the boundary. If it hits parentID == 0, it means no ancestor crossed the boundary, which should never happen, since the method is only called when a hit is created and some ancestor _must_ have crossed the boundary. If you prefer, I can make a `bool strict` function argument, so that an exception is thrown only if the function is executed in 'strict mode', and otherwise it throws just a warning?

Reply:
@tklijnsma , parentID=0 means that it is a primary track, so it is not a subject of exception or warning. From my point of view some other action is required. By the way, most part of primary tracks will path the boundary. Or I miss something?

Reply:
@civanch , This method `findBoundaryCrossingParent`, is specifically called for G4Tracks that are leaving a hit in a calorimeter. Clearly either the track itself or an ancestor _must_ have crossed the boundary into the calorimeter, otherwise a hit could not have been made. If during the ancestor lookup you reach parentID==0 (i.e. no further parent), something must have gone wrong with marking a track as 'crossed the boundary', hence the exception.

I think I understand the confusion: I wrote `findBoundaryCrossingParent` to be used _only_ for tracks that leave a hit, not just for any track.

Perhaps it could be renamed `findBoundaryCrossingParent` --> `findBoundaryCrossingParentForTrackThatLeavesAHit` (or something more concise)

Reply:
@tklijnsma, let us assume that primary muon or charged pion of high energy is tracked. These particles produce hits and they path the boundary of the calorimeter.  A muon has very high probability do not disappear before the boundary,  pion has some probability to have a decay or nuclear interaction but there is indeed a big probability that it will path the boundary. Why we should have exception at this point?

Reply:
@civanch the G4Track that is input to `findBoundaryCrossingParent` here is the G4Track that is _currently being processed by Geant_, not whatever the primary is. The whole point of finecalo is to assign hits to whatever ancestor crossed the boundary. The entire concept of primaries is not used for finecalo.

If a hit is created in a (fine) calorimeter, but no (ancestor of) the current G4Track entered into that calorimeter, it is an error. Something must enter the calorimeter to create hits in it. Note this all applies to finecalo, not regular CMSSW. None of this logic will affect how regular CMSSW treats primaries.

Reply:
In Geant4, G4Track, which has parentID=0 called primary. It is a normal G4Track produced at IR, this track may have hits in any sub-detector.

Reply:
Let's say a G4Tracks decay chain looks as follows:

Track 1 -> Track 100123 -> Track 100125

And let's say Track 1 crosses the boundary into the calorimeter. When Track 100125 creates a hit, `findBoundaryCrossingParent(Track 100125)` is called, which will look back into the decay chain until it finds a track that crossed the boundary. In this case it would first check Track 100123, check if it crosses the boundary (it doesn't), then check Track 1, which crosses the boundary, and the loop stops. The code happily returns Track 1 as the boundary-crossing parent.

Now let's say something is wrong with the boundary-crossing marking of tracks. The loop would first check if Track 100123 crosses the boundary (is doesn't), then Track 1 (it doesn't, because the marking is broken), and then it would find "Track 0". There are no further parents to check, so the hit cannot be assigned to any boundary-crossing parent. This is an error, because how can a hit be created in a detector if no track ever entered the calorimeter?

Reply:
@tklijnsma , I am sorry, now I understand that paretID==0 means only broken sequence in MCtruth mechanism.